### ISSUE TITLE
Masked Multi Head Attention

### ISSUE THREAD

**Issue by Luoxiaoyu828** [imperial-qore/TranAD#7]
```# Phase 1 - Without anomaly scores
c = torch.zeros_like(src)
x1 = self.fcn(self.transformer_decoder1(*self.encode(src, c, tgt)))   # x1-->O1
# Phase 2 - With anomaly scores
c = (x1 - src) ** 2
x2 = self.fcn(self.transformer_decoder2(*self.encode(src, c, tgt)))
```
Excuse me, in the TranAD model, which code is Masked Multi Head Attention reflected in? From the above code, the encoders of stages 1 and 2 are the same, and Masked Multi Head Attention is not involved in stage 2.

**Comment by shreshthtuli** [comment]
Please refer to the source code here https://pytorch.org/docs/stable/_modules/torch/nn/modules/transformer.html#TransformerDecoder

**Comment by Luoxiaoyu828** [comment]
Sorry, the code I have doubts is as follows:
https://github.com/imperial-qore/TranAD/blob/main/src/models.py
line 490--524
```
# Proposed Model + Self Conditioning + Adversarial + MAML (TKDE 21)
class TranAD(nn.Module):
	def __init__(self, feats):
		super(TranAD, self).__init__()
		self.name = 'TranAD'
		self.lr = lr
		self.batch = 128
		self.n_feats = feats
		self.n_window = 10
		self.n = self.n_feats * self.n_window
		self.pos_encoder = PositionalEncoding(2 * feats, 0.1, self.n_window)
		encoder_layers = TransformerEncoderLayer(d_model=2 * feats, nhead=feats, dim_feedforward=16, dropout=0.1)
		self.transformer_encoder = TransformerEncoder(encoder_layers, 1)
		decoder_layers1 = TransformerDecoderLayer(d_model=2 * feats, nhead=feats, dim_feedforward=16, dropout=0.1)
		self.transformer_decoder1 = TransformerDecoder(decoder_layers1, 1)
		decoder_layers2 = TransformerDecoderLayer(d_model=2 * feats, nhead=feats, dim_feedforward=16, dropout=0.1)
		self.transformer_decoder2 = TransformerDecoder(decoder_layers2, 1)
		self.fcn = nn.Sequential(nn.Linear(2 * feats, feats), nn.Sigmoid())

	def encode(self, src, c, tgt):
		src = torch.cat((src, c), dim=2)
		src = src * math.sqrt(self.n_feats)
		src = self.pos_encoder(src)
		memory = self.transformer_encoder(src)
		tgt = tgt.repeat(1, 1, 2)
		return tgt, memory

	def forward(self, src, tgt):
		# Phase 1 - Without anomaly scores
		c = torch.zeros_like(src)
		x1 = self.fcn(self.transformer_decoder1(*self.encode(src, c, tgt)))
		# Phase 2 - With anomaly scores
		c = (x1 - src) ** 2
		x2 = self.fcn(self.transformer_decoder2(*self.encode(src, c, tgt)))
		return x1, x2
```
## x2 = self.fcn(self.transformer_decoder2(*self.encode(src, c, tgt)))
for self.encode(src, c, tgt) 
What I understand is that Masked Multi Head Attention needs to be used in Phase 2.

**Comment by Luoxiaoyu828** [comment]
Thanks for your reply!
I combined the transformer code to sort out the network structure of TranAD.https://github.com/imperial-qore/TranAD/blob/main/src/models.py [line490-524], 
which is exactly as you said: "Masked Multi-Head Attention is ued in both phases as it is part of the decoder. It is implemented as part of the TransformerDecoder model in pytorch".

I have several new doubts in the process of understanding the code, if you could help me to solve it, I would be very grateful!

1. During the training process of the TranAD model https://github.com/imperial-qore/TranAD/blob/main/main.py[line 251-273], 
I found that when calculating the loss, the weights of the reconstruction loss and the adversarial loss are (1 / n)  and  (1 - 1/n), respectively. Will such changes affect the training results?

2. It is still the Masked Multi-Head Attention problem. In the `z = model(window, elem)`[line 264], I debugged and still did not find that there is masked data in the decoder process. This confuses me a lot.

I would appreciate your suggestions,
Cheers

**Comment by Luoxiaoyu828** [comment]
Thank you very much for your reply！

I will continue to study the code and look forward to your reply.

Thank you again！
